### PR TITLE
chore(observability): fully implement HTTP external resource for component validation

### PR DESCRIPTION
**Pull request by tobz**
This PR fully fleshes out the HTTP external resource used for component validation.

We've added support for the remaining two modes (pull input, and push output), and in doing so, have implemented `ValidatableComponent` for more components to exercise the external resource.

We've explicitly left the "pull output" mode unimplemented, which would handle pulling events from an HTTP server driven by a sink. As the `prometheus_exporter` sink is the only sink which operates in that mode, but requires special handling (1), I decided to leave it marked `todo!()` for now. We have the ability to implement it if the need arises.

1. As we need the ability to aggregate/deduplicate metrics emitted by the `prometheus_exporter` sink, we'll end up needing to implement a custom external resource to handle that particular sink. Either that, or we'll implement the pull output mode for the HTTP external resource, and add the ability to specify a custom batcher/handler/something to achieve the aggregation/deduplication bit.

### CODE REVIEW

File: src/components/validation/runner/io.rs [hunk 1]
```diff
@@ -89,16 +89,24 @@ impl InputEdge {
...
-            // TODO: Read events from `rx` and send them to the component topology via our Vector
-            // gRPC client that connects to the Vector source.
-            while let Some(_event) = rx.recv().await {}
+            while let Some(test_event) = rx.recv().await {
+                let request = PushEventsRequest {
+                    events: vec![test_event.into_event().into()],
+                };
+                match client.push_events(request).await {
+                    Ok(_response) => {}
+                    Err(_e) => {}
```
Comment (neuronull) [inline src/components/validation/runner/io.rs:107]: Is this a scenario that could happen? Just wondering if a logging the error would be worthwhile for debuggability.

File: src/components/validation/validators/component_spec.rs [hunk 2]
```diff
@@ -23,11 +23,60 @@ impl Validator for ComponentSpecValidator {
...
+                // got an output event for every valid input event, where, ostensibly, any event
+                // that wasn't marked for modification would represent a valid input event.
+                //
+                // Thinking it through, however, it's not 100% clear at this moment whether or not
+                // we'll ever have a test case, with an expectation of partial success, that
+                // intentionally modifies one or more input events but where some of those modified
+                // events will lead to output events.
+                //
+                // I might be overthinking it, but we'll leave this as-is for now because it's
+                // certainly the simplest check that matches reality.
```
Comment (neuronull) [inline src/components/validation/validators/component_spec.rs:69]: YAGNI approach sounds good here, thanks for calling it out in detail.

Is this really a TODO though? 💭

File: tests/validation/components/sources/http_client.yaml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+- name: happy path
+  expectation: success
+  events:
+    - simple message 1
+    - simple message 2
+    - simple message 3
+- name: sad path
```
Comment (neuronull) [inline tests/validation/components/sources/http_client.yaml:7]: 👍 much better than "error path" 😄

File: src/components/validation/resources/http.rs [hunk 4]
```diff
@@ -128,22 +224,180 @@ fn spawn_input_http_client(
...
+/// Spawns an HTTP client that pulls events by making requests to an HTTP server driven by a sink.
 #[allow(clippy::missing_const_for_fn)]
 fn spawn_output_http_client(
-    _config: HttpConfig,
+    _config: HttpResourceConfig,
     _codec: ResourceCodec,
     _output_tx: mpsc::Sender<Event>,
     _task_coordinator: &TaskCoordinator<Configuring>,
 ) {
+    // The `prometheus_exporter` sink is the only sink that exposes an HTTP server which must be
```
Comment (neuronull) [inline src/components/validation/resources/http.rs:295]: ?

```suggestion
    // TODO: The `prometheus_exporter` sink is the only sink that exposes an HTTP server which must be
```

**Review by neuronull — COMMENTED** [review]
Mostly clicks for me, thanks for documenting the flow and thought process.

File: src/components/validation/runner/io.rs (hunk 1 above)
Comment (tobz) [reply src/components/validation/runner/io.rs:107]: Yeah, fair point. I'll make it spit out an error log.

File: src/components/validation/validators/component_spec.rs (hunk 2 above)
Comment (tobz) [reply src/components/validation/validators/component_spec.rs:69]: Yeah, I'll transform it to a more succinct note somewhere above.

File: src/components/validation/runner/io.rs (hunk 1 above)
Comment (tobz) [reply src/components/validation/runner/io.rs]: Fixed in 386072039.

File: src/components/validation/validators/component_spec.rs (hunk 2 above)
Comment (tobz) [reply src/components/validation/validators/component_spec.rs]: Fixed in 386072039.

File: src/components/validation/resources/http.rs (hunk 4 above)
Comment (tobz) [reply src/components/validation/resources/http.rs]: Fixed in 386072039.

**Review by neuronull — APPROVED** [review]

File: src/components/validation/resources/http.rs [hunk 5]
```diff
@@ -128,22 +224,180 @@ fn spawn_input_http_client(
...
     _output_tx: mpsc::Sender<Event>,
     _task_coordinator: &TaskCoordinator<Configuring>,
 ) {
+    // TODO: The `prometheus_exporter` sink is the only sink that exposes an HTTP server which must be
+    // scraped... but since we need special logic to aggregate/deduplicate scraped metrics, we can't
+    // use this generically for that purpose.
+    todo!()
+}
+
+fn spawn_http_server<H, F, R>(
```
Comment (spencergilbert) [inline src/components/validation/resources/http.rs:301]: Missing a comment for this function

**Review by spencergilbert — APPROVED** [review]

File: src/components/validation/resources/http.rs [hunk 6]
```diff
@@ -1,9 +1,28 @@
...
+    net::{IpAddr, SocketAddr},
+    str::FromStr,
+    sync::Arc,
+};
+
+use axum::{
+    response::IntoResponse,
+    routing::{MethodFilter, MethodRouter},
+    Router,
+};
```
[the remark this answers is not in the document]
Comment (tobz) [reply src/components/validation/resources/http.rs:12]: 😈

File: src/components/validation/resources/http.rs (hunk 5 above)
Comment (tobz) [reply src/components/validation/resources/http.rs:301]: I/we usually only document public functions, though... 🥺

Comment (spencergilbert) [reply src/components/validation/resources/http.rs:301]: Oof, sorry - bad reading comprehension. I thought it was another in the line of documented functions above it. Now I see that it's called _in_ those.

**Pull request merged**